src/buzzer.rs

```diff
@@ -1,63 +1,90 @@
-pub struct Buzzer<'t> {
-    timer: Timer<'t, LowSpeed>,
-    channel: Channel<'t, LowSpeed>,
+#[derive(Debug)]
+pub enum Error {
+    FreqTooLow(u16),
+    DurationTooLongForFreq((u16, u16)),
+    StopError(esp_hal::rmt::Error),
+    TxError(esp_hal::rmt::Error),
 }
 
-impl<'t> Buzzer<'t> {
+#[derive(Debug)]
+enum Resource<'b> {
+    Channel(Channel<'b, Blocking, Tx>),
+    ContinuousTx(ContinuousTxTransaction<'b>),
+}
+
+/// Buzzer driver using RMT peripheral to generate the signal.
+#[derive(Debug)]
+pub struct Buzzer<'b> {
+    buf: [PulseCode; PULSE_COUNT + 1], // end marker required
+    resource: Resource<'b>,
+}
+
+impl<'b> Buzzer<'b> {
     /// Build a new buzzer driver.
     ///
-    /// This required a LEDC peripheral driver `ledc` and the output `pin`
-    pub fn new(mut ledc: Ledc<'t>, pin: AnyPin<'t>) -> Self {
-        // Initialize and create handle for LEDC peripheral
-        //let mut ledc = Ledc::new(led_perif);
-        ledc.set_global_slow_clock(LSGlobalClkSource::APBClk);
-
-        let channel = ledc.channel(channel::Number::Channel0, pin);
-        let timer = ledc.timer::<LowSpeed>(timer::Number::Timer0);
-        Buzzer { timer, channel }
+    /// This requires an RMT peripheral driver `rmt` and the output `pin`
+    pub fn new(rmt: Rmt<'b, Blocking>, pin: AnyPin<'b>) -> Self {
+        let channel = rmt
+            .channel0
+            .configure_tx(&TxChannelConfig::default()
+                          .with_clk_divider((ESP32S3_RMT_DEFAULT_CLK_FREQ_HZ / CLK_FREQ_HZ).try_into().unwrap())
+                          .with_idle_output_level(Level::Low)
+                          .with_idle_output(false)
+                          .with_carrier_modulation(false)
+                          .with_carrier_high(1)
+                          .with_carrier_low(1)
+                          .with_carrier_level(Level::Low)
+            ).expect("RMT channel0 could not configure")
+            .with_pin(pin);
+        Buzzer { buf: [PulseCode::default(), PulseCode::end_marker()], resource: Resource::Channel(channel) }
     }
 
-    // Private function to do the actual PWM frequency and duty settings.
-    fn configure(&'t mut self, freq: Rate, duty: u8) {
-        //let mut timer = self.ledc.timer::<LowSpeed>(timer::Number::Timer0);
-        //let mut channel = self.ledc.channel(channel::Number::Channel0, self.pin);
-
-        self.timer
-            .configure(timer::config::Config {
-                duty: timer::config::Duty::Duty5Bit,
-                clock_source: timer::LSClockSource::APBClk,
-                frequency: freq,
-            })
-            .unwrap();
-
-        self.channel
-            .configure(channel::config::Config {
-                timer: &self.timer,
-                duty_pct: duty,
-                drive_mode: DriveMode::PushPull,
-            })
-            .unwrap();
-    }
+    pub fn tone(mut self, freq_hz: u16, duration_ms: u16) -> Result<Self, (Self, Error)> {
```

Comment:
Why do you need to take ownership of `self` and then return `self` again ? is there something that prevents us using a mutable reference here ?